### PR TITLE
Move message header mode as a generic property

### LINKED ISSUE #408: Make message headers optional and configurable for incoming messages
Currently, `org.springframework.cloud.stream.binder.EmbeddedHeadersMessageConverter#extractHeaders` is invoked for `org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder` unconditionally, which makes integration with 3rd party message producer (i.e. non-Spring Cloud Stream) painful.

I suggest adding a configuration property which would allow omitting headers in an incoming message altogether.

I have a case when other application adds JSON-encoded `String` messages to the Redis list. These messages are successfully decoded and handled, but each message leads to error with the following stacktrace:

```java
2016-03-10 12:32:30.395 ERROR 89467 --- [hannel-adapter1] disMessageChannelBinder$ReceivingHandler : Could not convert message: 7B2269[HEX message omitted]6C7D

java.lang.StringIndexOutOfBoundsException: String index out of range: 523
	at java.lang.String.checkBounds(String.java:385) ~[na:1.8.0_72]
	at java.lang.String.<init>(String.java:425) ~[na:1.8.0_72]
	at org.springframework.cloud.stream.binder.EmbeddedHeadersMessageConverter.oldExtractHeaders(EmbeddedHeadersMessageConverter.java:134) ~[spring-cloud-stream-1.0.0.M4.jar:1.0.0.M4]
	at org.springframework.cloud.stream.binder.EmbeddedHeadersMessageConverter.extractHeaders(EmbeddedHeadersMessageConverter.java:104) ~[spring-cloud-stream-1.0.0.M4.jar:1.0.0.M4]
	at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$ReceivingHandler.handleRequestMessage(RedisMessageChannelBinder.java:362) ~[spring-cloud-stream-binder-redis-1.0.0.M4.jar:1.0.0.M4]
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:99) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:147) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:120) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$1.doWithRetry(RedisMessageChannelBinder.java:217) [spring-cloud-stream-binder-redis-1.0.0.M4.jar:1.0.0.M4]
	at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$1.doWithRetry(RedisMessageChannelBinder.java:213) [spring-cloud-stream-binder-redis-1.0.0.M4.jar:1.0.0.M4]
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:263) [spring-retry-1.1.2.RELEASE.jar:na]
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:168) [spring-retry-1.1.2.RELEASE.jar:na]
	at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1.doSend(RedisMessageChannelBinder.java:213) [spring-cloud-stream-binder-redis-1.0.0.M4.jar:1.0.0.M4]
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:392) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105) [spring-messaging-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:105) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.popMessageAndSend(RedisQueueMessageDrivenEndpoint.java:224) [spring-integration-redis-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.access$300(RedisQueueMessageDrivenEndpoint.java:54) [spring-integration-redis-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint$ListenerTask.run(RedisQueueMessageDrivenEndpoint.java:323) [spring-integration-redis-4.2.4.RELEASE.jar:na]
	at org.springframework.integration.util.ErrorHandlingTaskExecutor$1.run(ErrorHandlingTaskExecutor.java:55) [spring-integration-core-4.2.4.RELEASE.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_72]
```

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Both the producer and consumer properties have `HeaderMode`
 - Handle the case of embeddedHeaders and raw for both the Sending/ReceivingHandlers in Redis binder

This resolves #408

### CODE REVIEW

**Comment by ilayaperumalg** [comment]
This PR is submitted with the changes on top of the PR #409 and the original commit for this PR is: 348cac1

**Comment by ilayaperumalg** [comment]
Rebased the PR.

File: spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java [hunk 1]
```diff
@@ -626,13 +628,13 @@ protected void handleMessageInternal(Message<?> message) throws Exception {
 				targetPartition = roundRobin() % numberOfKafkaPartitions;
 			}
 
-			if (Mode.embeddedHeaders.equals(producerProperties.getMode())) {
+			if (HeaderMode.embeddedHeaders.equals(producerProperties.getHeaderMode())) {
```
Comment (mbogoevici) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java]: Wondering if we generalize this and move it to superclass. Rabbit may ignore.

Comment (ilayaperumalg) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java]: This PR actually does exactly that (generalising the header mode). Am I missing something?

**Comment by ilayaperumalg** [comment]
Rebased the PR.

File: spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java (hunk 1 above)
Comment (mbogoevici) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java]: I meant moving this to `AbstractBinder`. Seems generic enough not to require each implementation to handle it.

**Comment by mbogoevici** [comment]
Rebased, squashed, merged: https://github.com/spring-cloud/spring-cloud-stream/commit/502d275b414c7291c75ac52e772eb38023254c0b

**Pull request closed without merging**